drivers/video/fbdev/bcm2708_fb.c

```diff
@@ -649,15 +774,19 @@ static void bcm2708_fb_copyarea(struct fb_info *info,
 				const struct fb_copyarea *region)
 {
 	struct bcm2708_fb *fb = to_bcm2708(info);
-	struct bcm2708_dma_cb *cb = fb->cb_base;
+	struct bcm2708_fb_dev *fbdev = fb->fbdev;
+	struct bcm2708_dma_cb *cb = fbdev->cb_base;
 	int bytes_per_pixel = (info->var.bits_per_pixel + 7) >> 3;
 
 	/* Channel 0 supports larger bursts and is a bit faster */
-	int burst_size = (fb->dma_chan == 0) ? 8 : 2;
+	int burst_size = (fbdev->dma_chan == 0) ? 8 : 2;
 	int pixels = region->width * region->height;
 
-	/* Fallback to cfb_copyarea() if we don't like something */
-	if (bytes_per_pixel > 4 ||
+	/* If DMA is currently in use (ie being used on another FB), then
+	 * rather than wait for it to finish, just use the cfb_copyarea
+	 */
+	if (bcm_dma_is_busy(fbdev->dma_chan_base) ||
```

Comment:
This feels race-tastic.
You check that the DMA isn't busy here, but don't make it busy yourself until bcm_dma_start in dma_memcpy. Admittedly you do check there that the DMA isn't busy, but even there you have a potential race.

I think this really ought to have proper semaphore or mutex protection. How much slower is cfb_copyarea than waiting on dma_memcpy?

Reply:
No idea of the performance of the various options. This was a fairly clunky mechanism to try and prevent two DMA's being run at the same time. I'll have a think about semaphores, but this set of changes is already rearchitecting a load of code.